Backend/CarRentalApp/CarRentalApp/Services/Identity/UserService.cs

```diff
@@ -1,33 +1,248 @@
+        /// <exception cref="SharedException">Cannot specify client role without additional info.</exception>
+        private Task UpdateRolesAsync(User user, UserDTO userDTO)
+        {
+            if (userDTO.Roles.Count < 1)
+            {
+                throw new SharedException(
+                    ErrorTypes.Invalid,
+                    "Role changing failed",
+                    "User must have at least 1 role"
+                );
+            }
+
+            if (user.Roles.Select(role => role.Role).SequenceEqual(userDTO.Roles))
+            {
+                throw new SharedException(
+                    ErrorTypes.Conflict,
+                    "Role changing failed",
+                    "User already has specified roles"
+                );
+            }
+
+            if (userDTO.Roles.Intersect(UserRole.ClientRoles).Any())
+            {
+                ValidateAge(user.DateOfBirth);
+
+                if (user.DriverLicenseSerialNumber == null)
+                {
+                    throw new SharedException(
+                        ErrorTypes.Conflict,
+                        "Role changing failed",
+                        "User cannot become a client without specifying DriverLicenseSerialNumber"
+                    );
+                }
+            }
+
+            user.Roles = userDTO.Roles.Select(role => new UserRole() {Role = role}).ToList();
+
+            return _userDAO.UpdateUserAsync(user);
+        }
+
+        private User CreateFromDTO(UserRegistrationDTO userRegistrationDTO)
+        {
+            var user = _userMapper.Map<UserRegistrationDTO, User>(userRegistrationDTO);
+
+            user.Salt = _passwordService.GenerateSalt();
+            user.HashedPassword = _passwordService.DigestPassword(userRegistrationDTO.Password, user.Salt);
+
+            user.Roles = new List<UserRole>() {new UserRole() {Role = Roles.None}};
+
+            return user;
+        }
+
+        /// <exception cref="SharedException">Invalid user age.</exception>
+        private void ValidateAge(DateTime dateOfBirth)
+        {
+            var minimumAge = _clientRequirements.MinimumAge;
+            var criticalDate = DateTime.Now.AddYears(minimumAge);
+
+            if (criticalDate.CompareTo(dateOfBirth) > 0)
+            {
+                throw new SharedException(
+                    ErrorTypes.Conflict,
+                    "Client`s data does not meet requirements",
+                    $"Client`s age must be greater than {minimumAge - 1}"
+                );
+            }
+        }
+
+        /// <exception cref="SharedException">Incorrect username or password.</exception>
+        private void ValidatePassword(User existingUser, UserLoginDTO userLoginDTO)
+        {
+            if (!_passwordService.VerifyPassword(
+                    existingUser.HashedPassword,
+                    existingUser.Salt,
+                    userLoginDTO.Password
+                ))
```

Comment:
```suggestion
            if (
                !_passwordService.VerifyPassword(
                    existingUser.HashedPassword,
                    existingUser.Salt,
                    userLoginDTO.Password
                )
            )
```

I also recommend to use positive if statements
it is easier to read